### PR TITLE
Add hard-sigmoid and hard-swish activations

### PR DESCRIPTION
**Pull request by jamjamjon**
I am building a model which use hardsigmoid() activation, basically equals: relu6(x+3) / 6
More details at: [https://pytorch.org/docs/stable/generated/torch.nn.functional.hardsigmoid.html#torch.nn.functional.hardsigmoid](url)

And I found Candle hasn't provide it yet. So This PR is for that.

### CODE REVIEW

File: candle-nn/src/ops.rs [hunk 1]
```diff
@@ -44,6 +44,12 @@ pub fn sigmoid(xs: &Tensor) -> Result<Tensor> {
     (xs.neg()?.exp()? + 1.0)?.recip()
 }
 
+pub fn hard_sigmoid(xs: &Tensor) -> Result<Tensor> {
+    // TODO: Should we have a specialized op for this?
+    let six = Tensor::new(6f32, xs.device())?;
+    ((xs + 3.0) / six)?.clamp(0f32, 1f32)
```
Comment (LaurentMazare) [inline candle-nn/src/ops.rs:50]: I think this is hard-swish and not hard-sigmoid in the pytorch nomenclature.

File: candle-nn/src/activation.rs [hunk 2]
```diff
@@ -29,7 +31,9 @@ impl super::Module for Activation {
             Self::Relu6 => xs.clamp(0f32, 6f32),
             Self::Silu => crate::ops::silu(xs),
             Self::Sigmoid => crate::ops::sigmoid(xs),
+            Self::HardSigmoid => crate::ops::hard_sigmoid(xs),
             Self::Swish => xs * crate::ops::sigmoid(xs)?,
+            Self::HardSwish => xs * crate::ops::hard_sigmoid(xs)?,
```
Comment (LaurentMazare) [inline candle-nn/src/activation.rs:36]: Why does hard-swish point at hard-sigmoid? My understanding is that these should be two different functions.

Comment (jamjamjon) [reply candle-nn/src/activation.rs:36]: hard-switch = x * hard-sigmoid 
Details at: [https://pytorch.org/docs/stable/generated/torch.nn.functional.hardswish.html#torch.nn.functional.hardswish](url)

File: candle-nn/src/ops.rs (hunk 1 above)
Comment (jamjamjon) [reply candle-nn/src/ops.rs:50]: I add two activation, hard-sigmoid and hard-swish.
hard-sigmoid = relu6(x+3) / 6
hard-swish = x * hard-sigmoid

Comment (LaurentMazare) [reply candle-nn/src/ops.rs:50]: Oops, my bad, this is hard-sigmoid, can you replace the `six` with an actual `6`?

File: candle-nn/src/activation.rs (hunk 2 above)
Comment (LaurentMazare) [reply candle-nn/src/activation.rs:36]: Ah right, thanks for clarifying.

File: candle-nn/src/ops.rs (hunk 1 above)
Comment (jamjamjon) [reply candle-nn/src/ops.rs:50]: There will be an error:
error[E0277]: cannot divide `std::result::Result<candle_core::Tensor, candle_core::Error>` by `{float}`
  --> candle-nn/src/ops.rs:50:17
   |
50 |     ((xs + 3.0) / 6.0)?.clamp(0f32, 1f32)
   |                 ^ no implementation for `std::result::Result<candle_core::Tensor, candle_core::Error> / {float}`
   |
   = help: the trait `std::ops::Div<{float}>` is not implemented for `std::result::Result<candle_core::Tensor, candle_core::Error>`
   = help: the following other types implement trait `std::ops::Div<Rhs>`:
             <std::result::Result<B, candle_core::Error> as std::ops::Div<candle_core::Tensor>>
             <std::result::Result<B, candle_core::Error> as std::ops::Div<&candle_core::Tensor>>

For more information about this error, try `rustc --explain E0277`.

Comment (LaurentMazare) [reply candle-nn/src/ops.rs:50]: Try `((xs + 3.0)? / 6)?` instead?

Comment (jamjamjon) [reply candle-nn/src/ops.rs:50]: `((xs + 3.0)? / 6)?.clamp(0f32, 1f32)` still got an error:     
error[E0277]: the trait bound `{integer}: std::borrow::Borrow<candle_core::Tensor>` is not satisfied
  --> candle-nn/src/ops.rs:53:18
   |
53 |     ((xs + 3.0)? / 6)?.clamp(0f32, 1f32)
   |                  ^ the trait `std::borrow::Borrow<candle_core::Tensor>` is not implemented for `{integer}`
   |
   = note: required for `candle_core::Tensor` to implement `std::ops::Div<{integer}>`

For more information about this error, try `rustc --explain E0277`.
warning: `candle-nn` (lib) generated 1 warning
error: could not compile `candle-nn` (lib) due to previous error; 1 warning emitted

But, This will be Ok

Comment (LaurentMazare) [reply candle-nn/src/ops.rs:50]: Yes, should be 6.0 instead of 6 indeed. Please use `div` instead of `/` and should be all good.

Comment (jamjamjon) [reply candle-nn/src/ops.rs]: Ok, fixed it.

**Comment by LaurentMazare** [comment]
Merged, thanks for the PR and sorry for the confusion on the function names.

**Comment by jamjamjon** [comment]
My bad! I made a typo!  I need to have a rest indeed. Thanks your patience.
